lib/Utils/ContextAwareTypeConversion.h

```diff
@@ -0,0 +1,615 @@
+        });
+  }
+
+  /// Generate a wrapper for the given memory space conversion callback. The
+  /// callback may take any subclass of `Attribute` and the wrapper will check
+  /// for the target attribute to be of the expected class before calling the
+  /// callback.
+  template <typename T, typename A, typename FnT>
+  TypeAttributeConversionCallbackFn wrapTypeAttributeConversion(
+      FnT &&callback) const {
+    return [callback = std::forward<FnT>(callback)](
+               Type type, Attribute attr) -> AttributeConversionResult {
+      if (T derivedType = dyn_cast<T>(type)) {
+        if (A derivedAttr = dyn_cast_or_null<A>(attr))
+          return callback(derivedType, derivedAttr);
+      }
+      return AttributeConversionResult::na();
+    };
+  }
+
+  /// Register a memory space conversion, clearing caches.
+  void registerTypeAttributeConversion(
+      TypeAttributeConversionCallbackFn callback) {
+    typeAttributeConversions.emplace_back(std::move(callback));
+    // Clear type conversions in case a memory space is lingering inside.
+    cachedDirectConversions.clear();
+    cachedMultiConversions.clear();
+  }
+
+  /// The set of registered conversion functions.
+  SmallVector<ConversionCallbackFn, 4> conversions;
+
+  /// The list of registered materialization functions.
+  SmallVector<MaterializationCallbackFn, 2> argumentMaterializations;
+  SmallVector<MaterializationCallbackFn, 2> sourceMaterializations;
+  SmallVector<TargetMaterializationCallbackFn, 2> targetMaterializations;
+
+  /// The list of registered type attribute conversion functions.
+  SmallVector<TypeAttributeConversionCallbackFn, 2> typeAttributeConversions;
+
+  /// A set of cached conversions to avoid recomputing in the common case.
+  /// Direct 1-1 conversions are the most common, so this cache stores the
+  /// successful 1-1 conversions as well as all failed conversions.
+  mutable DenseMap<TypeAndAttribute, Type> cachedDirectConversions;
+  /// This cache stores the successful 1->N conversions, where N != 1.
+  mutable DenseMap<TypeAndAttribute, SmallVector<Type, 2>>
+      cachedMultiConversions;
+  /// A mutex used for cache access
+  mutable llvm::sys::SmartRWMutex<true> cacheMutex;
+};
+
+// An AttributeAwareTypeConverter for which the attribute is determined uniquely
+// by a specific string name on the defining op or as a func arg attr.
+struct UniquelyNamedAttributeAwareTypeConverter : ContextAwareTypeConverter {
+  using ContextAwareTypeConverter::ContextAwareTypeConverter;
+
+ public:
+  UniquelyNamedAttributeAwareTypeConverter(StringRef attrName)
+      : attrName(attrName) {}
+
+  FailureOr<Attribute> getContextualAttr(Value value) const override {
+    return findAttributeAssociatedWith(value, attrName);
+  }
+
+  LogicalResult convertFuncSignature(
+      FunctionOpInterface funcOp, SmallVectorImpl<Type> &newArgTypes,
+      SmallVectorImpl<Type> &newResultTypes) const override {
```

Comment:
Add a comment here saying the default impl does not handle function declaration and add a guard against func declaration here. Otherwise this will run into segfault.

Reply:
I believe this is fine. `tests/Dialect/Secret/Conversions/secret_to_ckks/ops.mlir` has a declaration-only func and it does not crash on this pattern. Despite not having SSA values for operands, func ops that are merely declarations still do have arg and result attrs (FunctionOpInterface requires these attributes are distinct ODS operands!). I was careful to ensure this default impl does not try to access any SSA values or block arguments.